cirq-core/cirq/transformers/target_gatesets/compilation_target_gateset.py

```diff
@@ -29,15 +29,44 @@
     import cirq
 
 
-def _create_transformer_with_kwargs(func: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
-    """Hack to capture additional keyword arguments to transformers while preserving mypy type."""
+def create_transformer_with_kwargs(transformer: 'cirq.TRANSFORMER', **kwargs) -> 'cirq.TRANSFORMER':
+    """Method to capture additional keyword arguments to transformers while preserving mypy type.
+
+    Returns a `cirq.TRANSFORMER` which, when called with a circuit and transformer context, is
+    equivalent to calling `transformer(circuit, context=context, **kwargs)`. It is often useful to
+    capture keyword arguments of a transformer before passing them as an argument to an API that
+    expects `cirq.TRANSFORMER`. For example:
+
+    >>> def run_transformers(transformers: List[cirq.TRANSFORMER]):
+    >>>     for transformer in transformers:
+    >>>         transformer(circuit, context=context)
+    >>>
+    >>> transformers: List[cirq.TRANSFORMER] = []
+    >>> transformers.append(
+    >>>     cirq.create_transformer_with_kwargs(
+    >>>         cirq.expand_composite, no_decomp=lambda op: cirq.num_qubits(op) <= 2
+    >>>     )
+    >>> )
+    >>> transformers.append(cirq.create_transformer_with_kwargs(cirq.merge_k_qubit_gates, k=2))
+    >>> run_transformers(transformers)
+
+
+    Args:
+         transformer: A `cirq.TRANSFORMER` for which additional kwargs should be captured.
+         **kwargs: The keyword arguments which should be captured and passed to `transformer`.
+
+    Returns:
+        A `cirq.TRANSFORMER` method `transformer_with_kwargs`, s.t. executing
+        `transformer_with_kwargs(circuit, context=context)` is equivalent to executing
+        `transformer(circuit, context=context, **kwargs)`.
+    """
 
-    def transformer(
+    def transformer_with_kwargs(
         circuit: 'cirq.AbstractCircuit', *, context: Optional['cirq.TransformerContext'] = None
     ) -> 'cirq.AbstractCircuit':
-        return func(circuit, context=context, **kwargs)  # type: ignore
+        return transformer(circuit, context=context, **kwargs)  # type: ignore
```

Comment:
This will raise a `SyntaxError` when kwargs contains "context". However, this will happen late - when the returned transformer is called - forcing someone to track down where it was instantiated. As a courtesy to the user, `create_transformer_with_kwargs` could check whether "context" is in kargs and raise immediately.

Reply:
Good point, I'll add this condition to the method.

Reply:
done